src/data.rs

```diff
@@ -0,0 +1,107 @@
+use core::fmt;
+
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct Reading {
+    temperature: f64,
+    humidity: f64,
+    pressure: i32,
+    dust_concentration: f64,
+    air_purity: String,
+}
+
+impl Reading {
+    pub(crate) fn new(
+        temperature: f64,
+        humidity: f64,
+        pressure: i32,
+        dust_concentration: f64,
+        air_purity: String,
+    ) -> Reading {
+        Reading {
+            temperature,
+            humidity,
+            pressure,
+            dust_concentration,
+            air_purity,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum AirPurity {
+    Dangerous,
+    High,
+    Low,
+    FreshAir,
+}
+
+impl AirPurity {
+    pub(crate) fn from_value(value: f64) -> AirPurity {
+        match value {
+            value if (f64::MIN..=50.0).contains(&value) => AirPurity::FreshAir,
+            value if value > 50.0 && value <= 100.0 => AirPurity::Low,
+            value if value > 100.0 && value <= 150.0 => AirPurity::High,
+            _ => AirPurity::Dangerous,
+        }
+    }
+}
+
+// implements fmt::Display for AirPurity so that we can call .to_string() on
+// each enum value
+impl fmt::Display for AirPurity {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            AirPurity::Low => write!(f, "Fresh Air"),
+            AirPurity::High => write!(f, "Low Pollution"),
+            AirPurity::Dangerous => write!(f, "High Pollution"),
+            AirPurity::FreshAir => write!(f, "Dangerous Pollution"),
+        }
+    }
+}
+
+pub(crate) fn random_gen_humidity() -> f64 {
```

Comment:
You could pass `rng` in as an argument for these functions so that you don't have to generate one for each function call. Then update the caller of these functions to invoke `thread_rng()` and pass in the value.